library/psa_crypto.c

```diff
@@ -673,13 +673,79 @@ static psa_algorithm_t psa_key_policy_algorithm_intersection(
+    if( PSA_ALG_IS_AEAD( alg1 ) && PSA_ALG_IS_AEAD( alg2 ) &&
+        ( PSA_ALG_AEAD_WITH_SHORTENED_TAG( alg1, 0 ) ==
+          PSA_ALG_AEAD_WITH_SHORTENED_TAG( alg2, 0 ) ) )
+    {
+        size_t alg1_len = PSA_ALG_AEAD_GET_TAG_LENGTH( alg1 );
+        size_t alg2_len = PSA_ALG_AEAD_GET_TAG_LENGTH( alg2 );
+        size_t max_len = alg1_len > alg2_len ? alg1_len : alg2_len;
+
+        /* If both are wildcards, return most restricitve wildcard */
+        if( ( ( alg1 & PSA_ALG_AEAD_MINIMUM_LENGTH_FLAG ) != 0 ) &&
+            ( ( alg2 & PSA_ALG_AEAD_MINIMUM_LENGTH_FLAG ) != 0 ) )
+        {
+            return( PSA_ALG_AEAD_WITH_AT_LEAST_THIS_LENGTH_TAG( alg1, max_len ) );
+        }
+        /* If only one is a wildcard, return specific algorithm if compatible. */
+        if( ( ( alg1 & PSA_ALG_AEAD_MINIMUM_LENGTH_FLAG ) != 0 ) &&
+            ( alg1_len <= alg2_len ) )
+        {
+            return( alg2 );
+        }
+        if( ( ( alg2 & PSA_ALG_AEAD_MINIMUM_LENGTH_FLAG ) != 0 ) &&
+            ( alg2_len <= alg1_len ) )
+        {
+            return( alg1 );
+        }
+    }
+    /* If the policies are from the same MAC family, check whether one
+     * of them is a minimum-MAC-length policy. Calculate the most
+     * restrictive tag length. */
+    if( PSA_ALG_IS_MAC( alg1 ) && PSA_ALG_IS_MAC( alg2 ) &&
+        ( PSA_ALG_FULL_LENGTH_MAC( alg1 ) ==
+          PSA_ALG_FULL_LENGTH_MAC( alg2 ) ) )
+    {
+        size_t alg1_len = PSA_MAC_TRUNCATED_LENGTH( alg1 );
+        size_t alg2_len = PSA_MAC_TRUNCATED_LENGTH( alg2 );
+        size_t max_len = alg1_len > alg2_len ? alg1_len : alg2_len;
+
+        /* If both are wildcards, return most restricitve wildcard */
+        if( ( ( alg1 & PSA_ALG_MAC_MINIMUM_LENGTH_FLAG ) != 0 ) &&
+            ( ( alg2 & PSA_ALG_MAC_MINIMUM_LENGTH_FLAG ) != 0 ) )
+        {
+            return( PSA_ALG_AT_LEAST_THIS_LENGTH_MAC( alg1, max_len ) );
+        }
+        /* If only one is a wildcard, return specific algorithm if compatible.
+         * Special case: specific MAC algorithm with '0' as length means full-
+         * length MAC, which is always allowed by a wildcard with the same
+         * base algorithm. */
+        if( ( ( alg1 & PSA_ALG_MAC_MINIMUM_LENGTH_FLAG ) != 0 ) &&
+            ( ( alg1_len <= alg2_len ) ||
+              ( alg2 == PSA_ALG_FULL_LENGTH_MAC( alg1 ) ) ) )
+        {
+            return( alg2 );
+        }
+        if( ( ( alg2 & PSA_ALG_MAC_MINIMUM_LENGTH_FLAG ) != 0 ) &&
+            ( ( alg2_len <= alg1_len ) ||
+              ( alg1 == PSA_ALG_FULL_LENGTH_MAC( alg2 ) ) ) )
+        {
+            return( alg1 );
+        }
+    }
     /* If the policies are incompatible, allow nothing. */
     return( 0 );
 }
 
 static int psa_key_algorithm_permits( psa_algorithm_t policy_alg,
                                       psa_algorithm_t requested_alg )
 {
+    /* A requested algorithm cannot be a wildcard. */
+    if( PSA_ALG_IS_WILDCARD( requested_alg ) )
```

Comment:
That's the issue with not having documentation for static functions. You're now arguing for a behaviour of `psa_key_algorithm_permits` that would permit a policy as 'requested_alg'.

Question: when you see a function called `psa_key_algorithm_permits`, would you assume it returns 'permitted' when you give it a policy of `PSA_ALG_AEAD_WITH_SHORTENED_TAG( GCM, 8)` and a requested algorithm of `PSA_ALG_AEAD_WITH_SHORTENED_TAG( GCM, 12 )`? Since there's no documentation, I had to go by the input names, and since one is explicitly a policy and the other is explicitly an algorithm, I don't feel a policy would be acceptable as requested algorithm.

Reply:
I'm not arguing that `psa_key_algorithm_permits` should necessarily permit a policy as `requested_alg`. I'm arguing that this change introduced a bug, a unit test caught it, but you changed the test rather than fix the bug.

I agree that this function should have documentation. Mind you `psa_key_policy_permits` just below is documented, but the documentation is not precise enough to cover this case. I probably didn't notice that this was a potentially problematic corner case when I wrote this function, or else I would have documented it.

We can either accept that `psa_key_algorithm_permits effectively implements the less-than relation on algorithm policies (and document it accordingly, now that we've identified the problem), or change it to doing (partial) validation on the algorithm. The former is easier but could make the code more complex if more complex wildcard policies are added. The latter requires changing this function from returning a boolean to returning an error code so that it can reject invalid algorithms with `INVALID_ARGUMENT`.